### ISSUE TITLE
[Question][Proposal] Potentially more performant and memory efficient Map impl

### ISSUE THREAD

**Issue by dadhi** [zio/zio#6722]
Hi, I am pretty new to Scala, though I am following what's happening in Zio for a couple of years now, from the trenches of the .NET stack.

There I have my own performant immutable Map implementation written in C# https://github.com/dadhi/ImTools#imtools. 
This year (2022) as an execise to learn Scala, I am converting the `ImHashMap` datastructure to `SMap` here: https://github.com/dadhi/SMap

But I am lacking the guidance and maybe more ambitious goal :-)

I wonder if ZIO would be interested and needed to have something like this under its sleeve?

**Comment by dadhi** [comment]
@adamgfraser Are there any Map involving benchmarks in Zio so I can test the alternative?

**Comment by adamgfraser** [comment]
@dadhi I think you could look at something like `ZSink.collectAllToMap`. One challenge is that in a lot of cases if you need more performance you can move to a mutable map or you really want a concurrent map.

**Comment by dadhi** [comment]
> One challenge is that in a lot of cases if you need more performance you can move to a mutable map or you really want a concurrent map.

Agree. Though imagine that you don't need this tradeoff and can have both immutable map properties (which is why I assume the map is selected for the task) and performance/less allocations. 
Moreover you may even tune some properties because you are in control. Here specifically, I am talking about using this custom Map for internal implementation rather than exposed API.
